src/plugins/data/public/query/query_string/dataset_service/dataset_service.ts

```diff
@@ -91,29 +92,54 @@ export class DatasetService {
     }
   }
 
-  public async cacheDataset(dataset: Dataset): Promise<void> {
-    const type = this.getType(dataset.type);
-    if (dataset && dataset.type !== DEFAULT_DATA.SET_TYPES.INDEX_PATTERN) {
-      const spec = {
-        id: dataset.id,
-        title: dataset.title,
-        timeFieldName: dataset.timeFieldName,
-        fields: await type?.fetchFields(dataset),
-        dataSourceRef: dataset.dataSource
-          ? {
-              id: dataset.dataSource.id!,
-              name: dataset.dataSource.title,
-              type: dataset.dataSource.type,
-            }
-          : undefined,
-      } as IndexPatternSpec;
-      const temporaryIndexPattern = await this.indexPatterns?.create(spec, true);
-      if (temporaryIndexPattern) {
-        this.indexPatterns?.saveToCache(dataset.id, temporaryIndexPattern);
+  public async cacheDataset(
+    dataset: Dataset,
+    services: Partial<IDataPluginServices>
+  ): Promise<void> {
+    const type = this.getType(dataset?.type);
+    try {
+      const asyncType = type && type.meta.isFieldLoadAsync;
+      if (dataset && dataset.type !== DEFAULT_DATA.SET_TYPES.INDEX_PATTERN) {
+        const fetchedFields = asyncType
+          ? await type?.fetchFields(dataset, services)
+          : ({} as IndexPatternFieldMap);
+        const spec = {
+          id: dataset.id,
+          title: dataset.title,
+          timeFieldName: dataset.timeFieldName,
+          fields: fetchedFields,
+          fieldsLoading: asyncType ? true : false,
+          dataSourceRef: dataset.dataSource
+            ? {
+                id: dataset.dataSource.id!,
+                name: dataset.dataSource.title,
+                type: dataset.dataSource.type,
+              }
+            : undefined,
+        } as IndexPatternSpec;
+        const temporaryIndexPattern = await this.indexPatterns?.create(spec, true);
+
+        // Load schema asynchronously if it's an async index pattern
+        if (asyncType && temporaryIndexPattern) {
```

Comment:
does async index patterns mean the metadata like field schema is not stored in the index pattern saved object?

Reply:
Async index patterns means that the fields are lazy loaded; i.e. the temporary index pattern is created with empty fields and when the fields are loaded in the index pattern gets hydrated with the fields.